### ISSUE TITLE
Web accessibility

### ISSUE THREAD

**Issue by andrew-edwards** [pacific-hake/hake-assessment#946]
(Notes for the record).

Kelli tried pdfmanagement package: "There has been a lot of development on the pdfmanagement package for accessibility but not to the extent that we need and maybe even some backwards movement." 

So adding in alt text by hand again in final version. `lualatex` works okay (we thought it might be needed); gives same results as `pdflatex` except for occasionally running credible intervals into the margins, fixed with 0934a050

@kellijohnson-NOAA - please add anything else of use. (This helps us for our DFO stuff also, plus an e-book we have to do some things on).

**Comment by cgrandin** [comment]
I made it so all figures, including the ones brought in as files are in the atl_fisg_text list and the CSV file:
c4553aefafd61772c623621420a1c97e46d0feb2

**Comment by kellijohnson-NOAA** [comment]
Here is a [tagged version](https://drive.google.com/file/d/1tJY0DbM1NNMJ0IoVL8Q4v3XGk63YAGFH/view?usp=sharing) that I am going to send in the morning unless you want to work on it some more @cgrandin? Tables are bad, but I think those are the only warnings that are present.

**Comment by cgrandin** [comment]
I looked at it more, and tried to manually make a table tagged but totally failed because the software won't even recognize some of the data in the table to create a tag, i.e. you can't even create a tag for every cell, which is required to have a properly tagged table. So I think it's impossible using Acrobat to fix these tables.

I forgot I wrote this Wiki last year:
https://github.com/pacific-hake/hake-assessment/wiki/How-to-make-the-document-web-accessible

The only thing we didn't do "properly" was to make the figure captions tags type <Caption>. Instead we have <P>. It is possible with lots of work (at least a day, probably more). Hopefully they are OK with this.

**Comment by kellijohnson-NOAA** [comment]
I was able to successfully compile the tagpdf.tex documentation file located [here](https://github.com/u-fischer/tagpdf/blob/main/doc/tagpdf.tex). Using lualatex on my local machine from a tinytex install. The resulting pdf was "compliant" right out of the gate and has Figures. Looks like tables are doable with work but not included in the example. Lots of this is still experimental and changing by the hour. A new version of LaTeX will be available this June that will have more functionality out of the gate. In short. We need to switch to using lualatex as  simple first step.

**Comment by andrew-edwards** [comment]
lualatex worked for me fine on Friday (from my miktex installation, but I'll switch to tinytex at some point), and gave identical pdf output. Thanks for this.... (Am currently doing web accessibility stuff for our e-book...).

**Comment by andrew-edwards** [comment]
Ann here tested the published version (on Adobe), and said:
"The hake assessment does have several untagged/partially tagged tables (eg. a, f, g, h - I only went that far). For ones that are tagged, the first column is mis-identified has a header row in most (easy to fix manually). For some that are partially tagged, several data cells are tagged together instead of individually. For something like table a, the tab panel identifies it as a table but when I expand the table the tag looks like this:  Figure and table captions should be tagged as captions. This is something that is an issue in Word as well though. We/CSAS always apply them manually in the PDF (pretty quick). "
Just for information. Interesting that some tables seem okay and some don't, so that could be something to look into for 2023.

**Comment by cgrandin** [comment]
@andrew-edwards thanks for having her take a look.

We (Kelli and I) identified all those things before submitting, so it's good to see someone else has the exact same take on it. It appears the NOAA compliance folks don't care about the table issues, although I suspect that will change in future years and we will be required to have 100% compliance so it's something to work on.

**Comment by kellijohnson-NOAA** [comment]
I think merged cells create the most problems.

**Comment by kellijohnson-NOAA** [comment]
Regarding merged cells ... tables like the following are really difficult to make accessible
![image](https://user-images.githubusercontent.com/4108564/213805986-c4ecdc88-5571-4cba-acf6-ccfff10fdcd0.png)
Can we talk about ways in which people would still be happy with the table but remove the headers of Spawning biomass and relative spawning biomass? Could we just use the mathematical symbols in each of the three columns?

**Comment by cgrandin** [comment]
Easy - just make two tables!

**Comment by cgrandin** [comment]
My last comment wasn't too helpful, I think it would be good to keep them in one table so your idea is good @kellijohnson-NOAA 

Tables b,c, and d in the Executive summary would have to be done.

This one might be impossible to change though:
![image](https://user-images.githubusercontent.com/4551051/216562678-ff7122f7-10c1-4585-adf9-8dd3a7311b5a.png)

The decision tables will be a nightmare for accessibility but we can't really change them.

**Comment by kellijohnson-NOAA** [comment]
For Table 23 ... all we have to do is put the cohort year abovt catch weight 000s t and the word cohort above M 000s t or wrap it using "[xxxx]\ncohort" and put it above the first cell for that group. It wouldn't be that hard, we just can no longer combine the columns into merged columns. The rest of the entries would just be blank for that row.

**Comment by cgrandin** [comment]
Ahh, sure that's a way to do it. I can do that.

Do you think the vertical lines in the table make it worse for tagging? I imagined it would because it's complex Latex

**Comment by kellijohnson-NOAA** [comment]
I am not sure about the lines but I can do an experiment and of a table with and without and see if the screen reader is any different. So, I will get back to you.

**Comment by cgrandin** [comment]
I didn't get to this. @kellijohnson-NOAA should we still do this this year?

**Comment by kellijohnson-NOAA** [comment]
Those tables you are referring to are in the executive summary and nothing was mentioned about them in the draft so we should just move on for this year and come back to the tagging beyond what we currently have next year.

**Comment by kellijohnson-NOAA** [comment]
Issues that I saw when looking at the final tagged version for 2023
* page 2: the number of pages in the citation has a link which is tagged, this doesn't need to be a link
* tables are tagged as paragraphs
* reading order is goofed up when figure is above a paragraph that started on a previous page
* tables that are too wide for the page are not tagged
* Tables e and f were not tagged at all
* Only the last three columns of table g were tagged, the multirow elements for the names in the first row make things really difficult to tag.
* figure j and the paragraph below it were not tagged
* reading order was non existent for pages 23-25
* long tables have identified columns but not rows and it was really difficult/impossible to tag the second page of the table
* figures tagged with tooltips do not have the same structure, i.e., figure then  paragraph, caption, paragraph so the alternative text is in a different, i.e., lower, level ... i do not actually know which one is "correct"
* headings are tagged as paragraphs not headings
* Figures and tables should always be at the end of a section in the executive summary otherwise we have to mess with the reading order

**Comment by cgrandin** [comment]
Thanks for the notes..

One thing it seems for sure we should move the figures and tables in the ES to the end of it.

Taking another look at the tagpdf documentation it looks like we can wrap section headers in html tags (easy to do) to have them tag correctly. 

It seems we can define our own structures using tagpdf. See section 4.3 of the CTAN doc. They even have FIG and TAB as examples. It may be possible to make custom tag structures for those difficult tables

**Comment by aaronmberger-nwfsc** [comment]
I think we will get push back moving all tables and figures to the end of the executive summary (maybe not, but I think we might).  It is so much easier to read when they are close[r] to the text.  I like the idea of having tables and figures at the end of each section in the executive summary.  It sounded like that could work and would be a nice compromise.

**Comment by cgrandin** [comment]
Ahh, I assumed they had to be moved to the end of the ES but if it’s just the end of the sections within it that’s even better.

Should we just agree to do that for 2024 and onwards ?

**Comment by kellijohnson-NOAA** [comment]
Here is a screen shot of the problem. On page 7 it will stop mid sentence from the previous page and read the table caption and table. Then it goes to (3) and finishes the paragraph. Then when transitioning from page 7 to 8 it again stops mid sentence and reads the alternative text for the figure and then the caption before continuing paragraph 3. I had to go in and manually change the reading order of every section in the executive summary.
![image](https://user-images.githubusercontent.com/4108564/219815737-326c8279-c688-41bb-815b-f6f77b6f541d.png)

**Comment by cgrandin** [comment]
Yep, that’s what I figured.
Apparently we could inject the tagpdf structures to manipulate reading order as well, same way we do alt text:
https://tex.stackexchange.com/questions/622742/how-to-set-reading-order-in-the-accessibility-pdf

**Comment by andrew-edwards** [comment]
Hopefully there's a way to keep the figures and tables within the text in the ES, as it really is much easier to read. 

Maybe there'll be Latex updates for next year, especially as Latex automatically does the placement (top, bottom, next page etc, [tbp]). Some sort of command around a paragraph to say to not break it up when tagging. But I appreciate I'm not the one doing this....

**Comment by cgrandin** [comment]
I incorporated the tagging structures (code) into the TeX file with the post-processor for the bookdown project (`post_process()` which is called from `hake_pdf()`).

I made all tables (except decision tables which we are limited by stakeholder and manager preference) only have one header row so this would significantly improve matters. I did rearrange the Cohort catch (see screenshot below) table to be in long form which we didn't discuss but will make accessibility much better and also allows for an unlimited number of cohorts in the table (we can leave 1999 in there for reference). We couldn't add more cohorts before unless we either dropped one or made the font tiny.

![image](https://github.com/pacific-hake/hake-assessment/assets/4551051/f35e0ead-2856-46e8-af48-dc1278393ce1)

I'll close it for now, as much of what was discussed is resolved and I think we should open a new issue if (when) we have more issues with accessibility.
